### PR TITLE
Functions of npm metadata generating for single npm package

### PR DESCRIPTION
**Pull request by yma88**
**Some steps tips for the npm metadata generating functions:**
Reorganize the npm content files structure from the input tarball as the input is a bit different from the normal npm contents structure:
Extract the input tarball, relocate the npm tarball file, move the version metadata package.json from the extracted files, then generate package metadata from the version metadata and write it into the proper location.

This covers two kinds of npm standard paths: single path and scoped path, it's more reliable to get the path from the version name meta than the analytic request path.

BTW, for some metadata fields types defined here, which I found differ from the ones in Indy npm, so I've followed the input samples. About the jinja2 template for json, it's a little complex to declare in template, looping deeply, I'll try to use the environment class to render the metas.

### CODE REVIEW

File: template/package.json.j2 [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  {%- endif %}],
+  "keywords": [
+  {%- if meta.keywords is defined %}
+    {% for key in meta.keywords -%}
+      "{{ key }}"{%- if not loop.last %},{%- endif %}
+    {% endfor %}
+  {%- endif %}],
+  "bugs": "{%- if meta.bugs is defined %}{{ meta.bugs }}{%- endif %}",
+  "homepage": "{%- if meta.homepage is defined %}{{ meta.homepage }}{%- endif %}"
+}
```
Comment (ligangty) [inline template/package.json.j2:78]: As python has native ability to generate json file, I'm thinking if this template is really needed. If we use the jinja template to generate package.json, we need to check the empty values with keys. For example, the last "homepage":"...", if the meta.homepage is not defined, I think the key "homepage" should also not be there. 
So @yma88  can you check if it is easy to generate the json file through the class object directly and if it is easy to do? If yes, we can do it instead and avoid this template I think

File: mrrc/metadata_version_npm.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    """Scan a file path and find version metadata
+    """
+    try:
+        with open(path) as version_meta_file:
+            version_meta_data = json.load(version_meta_file)
+        version_schema = marshmallow_dataclass.class_schema(NPMVersionMetadata)()
+        version = version_schema.load(version_meta_data)
+        return version
+    except ValidationError:
+        logging('Error: Failed to validate version metadata!')
```
Comment (ligangty) [inline mrrc/metadata_version_npm.py:101]: Python don't forbid two classes in one single file, so maybe we should organize them by module(Which is a single python file) but not by class(See https://stackoverflow.com/questions/106896/possibilities-for-python-classes-organized-across-files/107836#107836). I recommend to wrap these two npm classes and its relative functions in one single file.(maybe metadata_npm.py)
And maybe later we will wrap all metadata related functions together into a single file, but let's do this later.

**Review by ligangty — COMMENTED** [review]
Generally looks good to me except two comments.

**Comment by yma88** [comment]
@ligangty Thanks for your comments!

For the template or native json generates the file solution,

The class in python needs to be JSON serializable(default is not) when using marshmallow-dataclass(or json) to dump(s), but it only accepts the built-in types: https://docs.python.org/3.8/library/json.html#encoders-and-decoders. So we need to make the class inherited from JSONEncoder or implement the JSONEncoder.default() function during the dump. It will be complex to cover all kinds of data types in the json, so I use __dict__ member of the instance to transform the standard python dict(both for version and package metadata), they could be JSON serializable as the structure in classes are simple enough so far.
Some shortcomings when using the json to generate metadata file directly: it could not customize the data formats like hiding fields of empty values, while it could also be met when using the object mapper in JAVA, so this should not be a big deal from my perspective.
The benefit is obvious that we don't need to edit the template file for json anymore once some fields or fields types changes.

So I leave the updates in the new commit(also merge the two metadata .py), and let me know if there is any issue around here before merging, thanks!

File: tests/test_npm_meta.py [hunk 3]
```diff
@@ -3,17 +3,16 @@
 import shutil
 import marshmallow_dataclass
 import mrrc.archive as archive
-import mrrc.metadata_version_npm as npm_version
-import mrrc.metadata_package_npm as npm_package
-from mrrc.metadata_package_npm import NPMPackageMetadata
+import mrrc.metadata_npm as metadata_npm
+from mrrc.metadata_npm import NPMPackageMetadata
```
Comment (ligangty) [inline tests/test_npm_meta.py:7]: As python function can be import here, so I think you can directly use:
from mrrc.metadata_npm import NPMPackageMetadata, scan_for_version, gen_package_metadata_file

and just use the functions directly in your code below

**Review by ligangty — COMMENTED** [review]
Looks good to me except a small improvement

File: tests/test_npm_meta.py (hunk 3 above)
Comment (yma88) [reply tests/test_npm_meta.py]: Okay good, thank you

**Review by ligangty — APPROVED** [review]
LGTM

**Pull request merged**